api/agent/resource.go

```diff
@@ -292,18 +241,19 @@ func (a *resourceTracker) GetResourceToken(ctx context.Context, memory uint64, c
 		}
 		isWaiting = false
 
-		if ctx.Err() != nil {
-			c.L.Unlock()
-			return
+		// We are here because either context expired or we found a resource
+		var t ResourceToken
+		if a.isResourceAvailableLocked(memory, cpuQuota, isAsync) {
```

Comment:
ok, this is one of the loop invariants (had to pause for this one, but this seems better than checking the ctx, since if we got one before ctx timed out we can try to return it -- slightly nitty, but var assigned in loop wouldn't require double check (i checked for race but we have a lock so it's not, that being said, time under lock...))